### DISCUSSION TITLE
[react-markdown] Uncaught ReferenceError: process is not defined

### DISCUSSION THREAD

**Discussion by bmakan** [remarkjs/remark#845] [Q&A] [answered]
Using webpack 5 and `react-markdown v7` I keep getting `Uncaught ReferenceError: process is not defined` in the browser and the app keeps crashing. Tried several different solutions described in [StackOverflow](https://stackoverflow.com/questions/30239060/uncaught-referenceerror-process-is-not-defined) and none worked.

For a React library there's too much stuff being imported from node. At the very least there should be a guide or warning on how to set up bundlers if it's out of your control (e.g. the assert issue from https://github.com/remarkjs/remark/discussions/831). None of the other libraries I use gave me such headaches.

In the end, I had to downgrade to 6.x.x for the error to go away.

**Comment by wooorm** [comment, marked as answer]
Please explain in more detail what code you are using and what the error is

**Reply by bmakan** [reply]
I just import the lib Component and render it.

```js
import ReactMarkdown from 'react-markdown';

...
return <ReactMarkdown children="__Hello World!__" />;
...
```

Browser app crashes and theres this in console:
> Uncaught ReferenceError: process is not defined
    js util.js:109
    Webpack 41

**Reply by wooorm** [reply]
There is no `process` use in `react-markdown`, nor is there a `util.js` file. I’m also assuming that webpack explained the error message in more detail than these three lines? Please provide more information.

**Reply by bmakan** [reply]
There's nothing more. Webpack bundle is created without any issues. The error originates from `node-modules/util/util.js`.

**Reply by wooorm** [reply]
Are you using other tools than webpack?

**Reply by bmakan** [reply]
This is the line causing the issue:
```js
if (process.env.NODE_DEBUG) {...}
```

So, something is using the `util.js` and which is in turn accessing the node's `process` module. It should be an env variable, but I am unsuccessful in defining it in any way.

**Reply by wooorm** [reply]
`util` is a polyfill for a builtin node module. For some reason your build setup is polyfilling that, but not `process`.
Nothing `react-markdown` uses, uses `process.env.NODE_DEBUG`. React does, however.

**Reply by bmakan** [reply]
> Are you using other tools than webpack?

Typescript, babel, some webpack loaders. But the issue only started once I installed, imported and tried to render the `ReactMarkdown` component.

**Reply by bmakan** [reply]
I was able to get it to work by installing `process` as dev dependency and including it via `ProvidePlugin`. Weird thing is, I already tried this before. Maybe I had to stop and relaunch the webpack since I use the `--watch` option.

Install process as dev dependency:
`npm i -D process`

Update webpack.config:
```js
const { ProvidePlugin } = require('webpack');

...
// webpack.config
plugins: [
    new ProvidePlugin({
         'process/browser',
    }),
],
...
```
